### ISSUE TITLE
Concurrent POST networks/create results in "Another app is currently holding the xtables lock"

### ISSUE THREAD

**Issue by indygreg** [docker/docker#26111]
**Output of `docker version`:**

```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:33:38 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:33:38 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 147
Server Version: 1.12.1
Storage Driver: overlay2
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host bridge null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-34-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 28.01 GiB
Name: ubuntu-vm-main
ID: QLXW:RPE6:XA5D:VFSY:OF6O:2TRE:WL22:DP7Y:37PS:ANX3:HMFU:ZKUZ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: indygreg
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Ubuntu 16.04 running under Hyper-V (but Hyper-V shouldn't matter for this bug report).

**Steps to reproduce the issue:**
1. Make concurrent calls to `POST networks/create` to create multiple networks in rapid succession
2. Docker API fails reliably with complaints of xtables lock being held.

Here is short Python script to reproduce the issue:

```
import docker
import concurrent.futures as futures
import uuid

client = docker.Client()

fs = []
with futures.ThreadPoolExecutor(8) as pool:
    for i in range(100):
        network_name = str(uuid.uuid1())
        fs.append(pool.submit(client.create_network, network_name, driver='bridge'))

    # check results of futures here if you want
```

**Describe the results you received:**

The server log is littered with the following:

```
Handler for POST /v1.24/networks/create returned error: Failed to Setup IP tables: Unable to enable SKIP DNAT rule: Another app is currently holding the xtables lock; waiting (1s) for it to exit...\n (<nil>)
Handler for POST /v1.24/networks/create returned error: Failed to Setup IP tables: Unable to allow intercontainer communication: Another app is currently holding the xtables lock; waiting (1s) for it to exit...\n (<nil>)
```

**Describe the results you expected:**

I would expect network API calls to work under load. If it could take several seconds for the request to complete, I'd appreciate if the client could specify a timeout so it doesn't have to retry requests until they succeed.

**Additional information you deem important (e.g. issue happens only occasionally):**

I'm creating many short-lived containers and user-defined networks as part of a testing environment. Each test essentially creates its own isolated "cluster" of containers. This is all using the Remote API directly (via the Python client). Running multiple tests concurrently was generally working fine until we started using user-defined networks. It appears these APIs can't handle load as well as e.g. the container create/start/stop/delete APIs.

**Comment by indygreg** [comment]
Looking at server logs a bit more, this error about the xtables lock appears to surface when starting containers and when deleting user-defined networks.

```
Failed to allocate and map port 20559-20559: Error iptables FORWARD: Another app is currently holding the xtables lock; waiting (1s) for it to exit...\n
Handler for POST /v1.24/containers/11e51b96ea31a76291925b65d7d2e00d63aee1b75489d155ea3ff2499c2a7379/start returned error: driver failed programming external connectivity on endpoint nauseous_carson (14370a0b35f2244985b6638f0a7b28cff2e9ebd1975b2c869d31f1f869fcf31b): Error iptables FORWARD: Another app is currently holding the xtables lock; waiting (1s) for it to exit...\n
Failed to clean iptables rules for bridge network: unable to remove inter-network communication rule: Another app is currently holding the xtables lock; waiting (1s) for it to exit...\n (<nil>)
Failed to clean iptables rules for bridge network: unable to remove inter-network communication rule: Another app is currently holding the xtables lock; waiting (1s) for it to exit...\n (<nil>)
```

So essentially user-defined networks are unusable when multiple clients are managing them simultaneously.

**Comment by indygreg** [comment]
Here's a fun server log:

```
"Failed to clean iptables rules for bridge network: unable to remove inter-network communication rule: Another app is currently holding the xtables lock; waiting (1s) for it to exit...\nAnother app is currently holding the xtables lock; waiting (3s) for it to exit...\nAnother app is currently holding the xtables lock; waiting (5s) for it to exit...\nAnother app is currently holding the xtables lock; waiting (7s) for it to exit...\nAnother app is currently holding the xtables lock; waiting (9s) for it to exit...\nAnother app is currently holding the xtables lock; waiting (11s) for it to exit...\nAnother app is currently holding the xtables lock; waiting (13s) for it to exit...\nAnother app is currently holding the xtables lock; waiting (15s) for it to exit...\nAnother app is currently holding the xtables lock; waiting (17s) for it to exit...\nAnother app is currently holding the xtables lock; waiting (19s) for it to exit...\nAnother app is currently holding the xtables lock; waiting (21s) for it to exit...\nAnother app is currently holding the xtables lock; waiting (23s) for it to exit...\nAnother app is currently holding the xtables lock; waiting (25s) for it to exit...\nAnother app is currently holding the xtables lock; waiting (27s) for it to exit...\n (<nil>)
```

Maybe this points to an underlying slowness in interaction with xtables?

**Comment by cpuguy83** [comment]
Thanks for the detailed report.

**Comment by dgageot** [comment]
@cpuguy83 For what it's worth, I was also able to reproduce this issue on D4W, with the [Dockercon voting app](https://github.com/docker/example-voting-app) by running these commands in a loop:

```
set -e
docker-compose build
docker-compose pull
docker-compose up -d
docker ps
docker-compose down
```

After 23 run, it failed with `Cannot start service result: driver failed programming external connectivity on endpoint votingapp_result_1 (90a32a833bcd37e0371c483aca656105bb0994e20b025c8dff447b3d28d3f56b): Error iptables FORWARD: Another app is currently holding the xtables lock; waiting (1s) for it to exit...'`

**Comment by justincormack** [comment]
@aboch any idea on this - it is coming up in our user reports eg https://github.com/docker/for-mac/issues/285 and CI so it seems quite prevalent.

**Comment by justincormack** [comment]
Oh I see it is assigned to @sanimej - any update?

**Comment by aboch** [comment]
This is a failure of the poor man logic in https://github.com/docker/libnetwork/blob/master/iptables/iptables.go#L405

In iptables 1.6 `bool xtables_lock(bool wait)` slightly modified the message it writes to stderr compared to previous versions.

https://git.netfilter.org/iptables/tree/iptables/xshared.c?h=v1.4.21:
```
fprintf(stderr, "Another app is currently holding the xtables lock; "
				"waiting for it to exit...\n");
```

https://git.netfilter.org/iptables/tree/iptables/xshared.c?h=v1.6.0:
```
fprintf(stderr, "Another app is currently holding the xtables lock; "
				"waiting (%ds) for it to exit...\n", waited);
```

**Comment by aboch** [comment]
Fixed by #28975
